### DISCUSSION TITLE
Should I polyfill ES6 globals?

### DISCUSSION THREAD

**Discussion by Cormanz** [caoccao/Javet#54] [Q&A] [answered]
Hello! So, I was doing some testing with Javet, trying to test out using threads with Javet. It almost worked, except I got an error like so, `JavetException: Runtime is already registered`, which is a topic for another time (if this is unintentional behaviour, I can make an issue)

I thought this might be because of garbage collection due to the runtime being stopped or halted or something like that, so I thought I'd make a `setInterval` to keep it alive, but I found out that didn't exist (in ES6) mode, and neither did `fetch` or `setTimeout`, but `Promise` and `Array` did exist though.

Should I polyfill these manually, or does V8 support these under some flag/option?

**Comment by caoccao** [comment]
> Hello! So, I was doing some testing with Javet, trying to test out using threads with Javet. It almost worked, except I got an error like so, `JavetException: Runtime is already registered`, which is a topic for another time (if this is unintentional behaviour, I can make an issue)

This exception shows your application shares V8Value among multiple V8Runtime instances which is prohibited. Javet actively validates that, otherwise, V8 and JVM will crash immediately.

> I thought this might be because of garbage collection due to the runtime being stopped or halted or something like that, so I thought I'd make a `setInterval` to keep it alive, but I found out that didn't exist (in ES6) mode, and neither did `fetch` or `setTimeout`, but `Promise` and `Array` did exist though.

Authentic V8 comes with a price that application needs to implement some API by itself, e.g. setTimeout, setInterval, fetch. And that is also one of the reasons that V8 mode is much more secure. Chrome has its own implementation as a web browser. Node.js has its own implementation upon libuv as a lightweight server.

Javet leaves that with Javet users for 2 major reasons.

1. Every application's usage is unique.
2. Certain implementation relies on certain library which is against the initial Javet design objective **Javet does not rely on any other libraries except for JDK, V8, Node.js**.

Most of Javet users just turn to Node.js to get the Node.js ecosystem for free, of course, leaving security a little bit behind.

> Should I polyfill these manually, or does V8 support these under some flag/option?

There is no flag for those features in V8. Do it by yourself. But, if your implementation only depends on JDK, please consider open source it along with Javet. Thank you.

**Reply by Cormanz** [reply]
> This exception shows your application shares V8Value among multiple V8Runtime instances which is prohibited. Javet actively validates that, otherwise, V8 and JVM will crash immediately.

That's very weird, since I don't ever make another runtime. When I run `func.call(null)` normally, func being a `V8ValueFunction`, it works fine, but then when I run it in a different thread, the same exact `func`, it throws that error for some reason.

The rest makes sense, I'll try to implement some of that myself. Thank you!

**Reply by caoccao** [reply]
Could you leave your minimal reproduceable code snippet?

And I created [Polyfill V8 Mode](https://github.com/caoccao/Javet/milestone/7) as a long-term milestone.

**Comment by caoccao** [comment, marked as answer]
Polyfilling V8 Mode is now at [Javenode](https://github.com/caoccao/Javenode). Please feel free to contribute or file issues there.
